client/components/swipeable/index.js

```diff
@@ -0,0 +1,296 @@
+}
+
+export const Swipeable = ( {
+	hasDynamicHeight = false,
+	children,
+	currentPage = 0,
+	onPageSelect,
+	pageClassName,
+	containerClassName,
+	...otherProps
+} ) => {
+	const [ pageWidth, setPageWidth ] = useState( null );
+	const [ swipeableArea, setSwipeableArea ] = useState();
+	const isRtl = useRtl();
+
+	const [ pagesStyle, setPagesStyle ] = useState( {
+		transform: `translate3d(0px, 0px, 0px)`,
+		transitionDuration: `300ms`,
+	} );
+	const [ dragStartData, setDragStartData ] = useState( null );
+
+	const pagesRef = useRef();
+	const resizeObserverRef = useRef();
+	const numPages = Children.count( children );
+
+	const getWidth = useCallback( () => {
+		return resizeObserverRef.current?.getBoundingClientRect().width;
+	}, [ resizeObserverRef ] );
+
+	const getPagesWidth = useCallback( () => {
+		if ( ! pageWidth ) {
+			return null;
+		}
+		return pageWidth * numPages;
+	}, [ pageWidth, numPages ] );
+
+	const getOffset = useCallback(
+		( index ) => {
+			const offset = pageWidth * index;
+			return isRtl ? offset : -offset;
+		},
+		[ isRtl, pageWidth ]
+	);
+
+	const updateEnabled = hasDynamicHeight && numPages > 1;
+
+	useUpdateLayout( updateEnabled, currentPage, ( hasTransitionDuration = true ) => {
+		const offset = getOffset( currentPage );
+		const transform = {
+			transform: `translate3d(${ offset }px, 0px, 0px)`,
+			transitionDuration: hasTransitionDuration ? `300ms` : `0ms`,
+		};
+
+		let height = { height: null };
+		const targetHeight = pagesRef.current?.querySelector( '.is-current' )?.offsetHeight;
+
+		if ( targetHeight && pagesStyle?.height !== targetHeight ) {
+			height = { height: null };
+		}
+		setPageWidth( getWidth() );
+		setSwipeableArea( pagesRef.current?.getBoundingClientRect() );
+		setPagesStyle( { ...pagesStyle, ...height, ...transform } );
+	} );
+
+	useEffect( () => {
+		if ( currentPage >= numPages ) {
+			onPageSelect( numPages - 1 );
+		}
+	}, [ numPages, currentPage, onPageSelect ] );
+
+	const handleDragStart = useCallback(
+		( event ) => {
+			const position = getDragPositionAndTime( event );
+			setDragStartData( position );
+			setPagesStyle( { ...pagesStyle, transitionDuration: `0ms` } ); // Set transition Duration to 0 for smooth dragging.
+		},
+		[ pagesStyle ]
+	);
+
+	const hasSwipedToNextPage = useCallback( ( delta ) => ( isRtl ? delta > 0 : delta < 0 ), [
+		isRtl,
+	] );
+	const hasSwipedToPreviousPage = useCallback( ( delta ) => ( isRtl ? delta < 0 : delta > 0 ), [
+		isRtl,
+	] );
+
+	const handleDragEnd = useCallback(
+		( event ) => {
+			if ( ! dragStartData ) {
+				return; // End early if we are not dragging any more.
+			}
+
+			const dragPosition = getDragPositionAndTime( event );
+			const delta = dragPosition.x - dragStartData.x;
+			const absoluteDelta = Math.abs( delta );
+			const velocity = absoluteDelta / ( dragPosition.timeStamp - dragStartData.timeStamp );
+
+			const hasMetThreshold = absoluteDelta > OFFSET_THRESHOLD || velocity > VELOCITY_THRESHOLD;
```

Comment:
I'm not sure but I think it might be better to make `OFFSET_THRESHOLD` be about 50% of the prev/next page width.

Reply:
> I think it might be better to make OFFSET_THRESHOLD be about 40% ~ 50% of the prev/next page width.

Can you explain a bit more? 

The intend right now is unless you scroll far enough then will eventually trigger the slide.

Reply:
I mean lots of carousels libarry use the percentage as the threshold. If the scroll percentage doesn't over about 40%~50%, it won't switch to the prev/next page. But now we use fixed 100px for that. it might be slightly weird if the width of the component is a bigger value. You can imagine that you just scroll a little or accidentally, but it switches to prev/next page immediately.

Reply:
Thanks for the explanation @arthur791004. Will see how I can adjust that part.